### ISSUE TITLE
Consider hosting plans for smaller services

### ISSUE THREAD

**Issue by pnorman** [openstreetmap/operations#287]
We've got a number of smaller services that don't occupy a great deal of CPU, RAM, or disk. These are mainly tertiary services like OTRS, IRC, OSMF site, etc. We've been using HP DL360 machines for this for some time now, putting multiple services on machines, but is this still what makes the most sense?

Based on the costs of #282, the G9 machines cost us approximately £700/yr, split halfway between hosting costs (power, space, internet) and amortized machine costs over how long they've lasted for us in the past. I'm not really worried about this cost, I care more about stability and what results in the lowest load on sysadmins.

I know @gravitystorm has some thoughts on this which would start with #169 and build from there.

Any changes are likely to be long-term and shouldn't weigh heavily on purchasing decisions, because machines will still need refreshing and we should be able to find a use for refreshed machines regardless.

**Comment by pnorman** [comment]
I'm not considering the main DB servers or large DB-based data consumers like nominatim and rendering servers - those have a different set of requirements, are a core part of what we do, and we did some rough evaluations of costs earlier this year.

**Comment by gravitystorm** [comment]
> I know @gravitystorm has some thoughts on this

Yes, I definitely have strong thoughts on this! And you are right, the first step in the long-term plan is to have a resilient storage option, so that we can move the services around between machines without losing state.

In the short term, there's a lot of these services that don't need to be run on OSMF-owned dedicated hardware. Also, we've recently seen how even well-run services interfere with each other when run on shared linux environments (e.g. OSQA blocking system upgrades for other services), so running each service independently might have some advantages too. 

Perhaps moving them to hosted VMs (e.g. on Hetzner) is a cheap and easy alternative? Although even with the smallest VMs, they would be more resourced (i.e. more expensive) than many of these services need. Another alternative would be to run them in VMs on hosted hardware, but that trades cost efficiency for extra sysadmin work in maintaining a VM-based platform.

£700 GBP per year would get us around 11 x 2CPU/4GB/40GB VMs from Hetzner, and there's a smaller option too if particular services don't need that much. Or there's a range of dedicated hardware options if we want to rent the hardware and run the VMs ourselves, that removes the overheads of having to organise and own the hardware ourselves.

Of course, I'd prefer that we were doing something more along the lines of resilient storage and disposable containers, but that's taking much longer than I expected to make any progress. So these options are hopefully less work but still useful in the interim.

**Comment by pnorman** [comment]
> £700 GBP per year would get us around 11 x 2CPU/4GB/40GB VMs from Hetzner, and there's a smaller option too if particular services don't need that much.

Note that the costs were per machine per year.

> Of course, I'd prefer that we were doing something more along the lines of resilient storage and disposable containers, but that's taking much longer than I expected to make any progress. So these options are hopefully less work but still useful in the interim.

I'd still like to go with resilient storage. I see that as incrementally useful with our current setup. I don't think renting the hardware would be incrementally useful because we'd be swapping managing hardware, which we do, with managing rentals, which we don't have the experience in.

Let's say we had the resilient storage working. How much work are containers on top of that?

**Comment by mboeringa** [comment]
Although this is probably a bit of a side step for this issue, I just wanted to share some experiences I have had with running an Ubuntu guest with PostgreSQL on an Oracle Virtualbox instance with Windows as host over the last 1.5 year or so.

I have taken this machine from Virtualbox 5.x to now 6.0.8 and going from PostgreSQL 9.6 to 11.3 + PostGIS 2.5.2 and running osm2pgsql 0.96 (but no diff updates) and JIT activated in PostgreSQL.

I also upgraded from Ubuntu 16.04 to 18.04 LTS, but one thing that continuously kept irritating me was the sluggish performance of the Gnome shell, which doesn't seem to be a particular VM issue, but a more general issue in Ubuntu based on what I read on the internet. Simply clicking icons to open e.g. a Terminal window was a non optimal experience to say the least.

More distressing though, was that under high CPU loads (near 100% CPU usage on all cores of a hyperthreaded 4 core i7), and high IO (> 125MB/s from a SATA SSD), database connected software like pgAdmin 4 stopped updating Dashboard graphs and caused connection issues, even though the main processes / queries that triggered the high load would finish if left running.

Most distressing though, was the inability to access the VM under such loads: the Gnome shell would become totally black and unresponsive, not even allowing you to login after trying to activate the VMs window by giving it focus.

However, after reading about Ubuntu 18.10 and 19.04 recently, that was supposed to alleviate such issues with the Gnome shell, I finally took the plunge and upgraded first to 18.10, and now also 19.04 which has a new updated kernel.

_This has made a huge difference!_

Finally, it is possible to keep using the VM under high loads. No more issues logging in, or opening Terminal windows or something like LibreOffice Writer while PostgreSQL is stressed in the background processing some complex SQL statement. I won't say it is always as snappy as I maybe would have liked, but the difference is as said huge and for the better, and totally workable.

I therefor recommend to run Ubuntu 18.10 minimum if possible on a VM like VirtualBox.
